client/with_dry_runs.go

```diff
@@ -0,0 +1,236 @@
+package client
+
+import (
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/mysteriumnetwork/payments/bindings"
+	"github.com/pkg/errors"
+)
+
+// WithDryRuns forces a dry run before running a write transaction on blockchain.
+// Ethereum client will perform a dry run on a transaction with no gas limit set.
+// This component will perform a dry run if and only if the gas limit is set to a non zero value.
+// In this way, the dry run is always performed before sending the transaction to the network.
+// For convenience, this component proxies read only calls to the underlying blockchain.
+type WithDryRuns struct {
+	bc blockchain
+}
+
+// NewWithDryRuns creates a new instance of client with dry runs.
+func NewWithDryRuns(bc blockchain) *WithDryRuns {
+	return &WithDryRuns{
+		bc: bc,
+	}
+}
+
+type gasLimitProvider interface {
+	GetGasLimit() uint64
+}
+
+func (cwdr *WithDryRuns) dryRun(glp gasLimitProvider, binding string, from, to common.Address, method string, params ...interface{}) (uint64, error) {
+	// If the gas limit is set to 0, ethereum client will do the estimation for us.
+	// We only force the estimation if the gas limit is set to a non zero value.
+	if glp.GetGasLimit() == 0 {
+		return 0, nil
+	}
+	parsed, err := abi.JSON(strings.NewReader(binding))
+	if err != nil {
+		return 0, errors.Wrap(err, "could not deserialize ABI")
+	}
+	input, err := parsed.Pack(method, params...)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not pack input")
+	}
+	msg := ethereum.CallMsg{From: from, To: &to, Data: input}
+	gas, err := cwdr.bc.EstimateGas(msg)
+	return gas, errors.Wrap(err, "could not estimate gas")
+}
+
+// TransferMyst transfers myst
+func (cwdr *WithDryRuns) TransferMyst(req TransferRequest) (tx *types.Transaction, err error) {
+	_, err = cwdr.dryRun(
+		req,
+		bindings.MystTokenABI,
+		req.Identity,
+		req.MystAddress,
+		"transfer",
+		req.Recipient,
+		req.Amount,
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "transaction dry run failed")
+	}
+
+	return cwdr.bc.TransferMyst(req)
+}
+
+// SettlePromise is settling the given consumer issued promise
+func (cwdr *WithDryRuns) SettlePromise(req SettleRequest) (*types.Transaction, error) {
+	rBytes := [32]byte{}
+	copy(rBytes[:], req.Promise.R)
+	lock := rBytes
```

Comment:
Why is this new lock variable needed? Could probably rename rBytes to lock and remove this lock var.

Reply:
it isn't, removed